src/hotspot/share/utilities/exceptions.cpp

```diff
@@ -262,8 +263,32 @@ void Exceptions::fthrow(JavaThread* thread, const char* file, int line, Symbol*
   va_list ap;
   va_start(ap, format);
   char msg[max_msg_size];
-  os::vsnprintf(msg, max_msg_size, format, ap);
+  int ret = os::vsnprintf(msg, max_msg_size, format, ap);
   va_end(ap);
+
+  // If ret == -1 then either there was a format conversion error, or the required buffer size
+  // exceeds INT_MAX and so couldn't be returned (undocumented behaviour of vsnprintf). Depending
+  // on the platform the buffer may be filled to its capacity (Linux), filled to the conversion
+  // that encountered the overflow (macOS), or is empty (Windows), so it is possible we
+  // have a truncated UTF-8 sequence. Similarly, if the buffer was too small and ret >= max_msg_size
+  // we may also have a truncated UTF-8 sequence. In such cases we need to fix the buffer so the UTF-8
+  // sequence is valid.
+  if (ret == -1 || ret >= max_msg_size) {
+    int len = (int) strlen(msg);
+    if (len > 0) {
```

Comment:
`truncate` asserts that len>5, you might need to adjust that.

Reply:
We know we only got here because the message was either huge (-1) or > 1K. We only get a zero length if we got -1 and are on Windows. Any length < max_msg_size means we got -1 and are on macOS and we have truncated prior to the conversion that caused the INT_MAX overflow. In theory it could be a single "%s" format but in practice we don't call fthrow that way. Also if we got the -1 then there are actually very few circumstances that can get us to this point because it needs to be an exception message that can relate to huge strings (which at the moment is trying to look up a class with an illegally long name - something that will soon be handled on the Java side before we get to the VM.)

So if we somehow were to trigger the len>5 assert, that is fine as it indicates something unusual/unexpected that we want to catch.